spirv_cross.hpp

```diff
@@ -856,6 +856,13 @@ class Compiler
 	bool instruction_to_result_type(uint32_t &result_type, uint32_t &result_id, spv::Op op, const uint32_t *args,
 	                                uint32_t length);
 
+	Bitset combined_decoration_for_member(const SPIRType &type, uint32_t index) const;
+	std::string type_to_string(const SPIRType &type) const;
```

Comment:
Isn't this basically type_to_glsl? I think it would make more sense if this is virtual and the other backends start using type_to_string instead. The old naming of type_to_glsl is a bit awkward to begin with since HLSL/MSL are also using it.

Right now it seems like a helper function specific to GLSL/Reflection, which is a bit weird.

Reply:
I concur with avoiding leaking, but I was unable to convert this to a static method on my first pass because some of the code paths use the internal `meta` array to do lookups.  I'll take a look again and see if there's not a way to push this into some kind GLSL helper library of static functions.

Reply:
I'm also considering that maybe trying to reduce code duplication here at all costs might be counterproductive. Clearly something is off if we need to contort the structure like this ... What do you think?

Reply:
The naming scheme is basically GLSL, but without any validation for ES or extensions. I think it is okay here to just implement the Reflection backends naming convention directly into CompilerReflection. Finding a clean way with standalone functions is fine too if you figure it out.

Reply:
If the desire is to keep `Compiler` clean of GLSL conventions, but that the JSON reflection _should_ use GLSL naming conventions (for simplicity and for lack of a better alternative) then perhaps I should revert the `CompilerReflection` class to deriving from `CompilerGLSL`.  That doesn't feel ideal either, though.  It seems more like what's needed is composition, rather than inheritance, where the internal representation of the SPIRV is in one class, and the conversion of that to the syntax of a given shader language is in a separate class hierarchy, of which a single concrete instance is owned by the `Compiler`.  This might simplify the Compiler hierarchy _and_ allow the JSON reflection to be customized so that the core types were expressed in terms of a specific target language (i.e. have a type be `float4` instead of `vec4` etc...).

However, this seems like a _big_ architectural change, at least for someone like me who isn't as familiar with the whole code base.  I think for the time being it would probably be simpler if I just reverted to having `CompilerReflection` derive from `CompilerGLSL`.

Reply:
Ye, that works too, CompilerReflection can inherit CompilerGLSL, that is the easiest, least invasive way to get this through. We can revisit later if we need non-GLSL centric reflection output.

Sorry for the back-and-forth. New backends are always a "big deal".

Reply:
It's no problem.  I find it pretty exciting to be able to contribute something to the project.